polynote-kernel/src/main/scala/polynote/kernel/util/NotebookContext.scala

```diff
@@ -0,0 +1,155 @@
+package polynote.kernel.util
+
+import cats.effect.{Concurrent, IO}
+import polynote.kernel.ResultValue
+import polynote.messages.{CellID, Notebook}
+
+import scala.annotation.tailrec
+
+/**
+  * Tracks all the CellContexts for a notebook. A good old-fashioned linked list in reverse, with an extra reference to the
+  * tip for easy append (i.e. predefs)
+  */
+class NotebookContext(implicit concurrent: Concurrent[IO]) {
+
+  @volatile private var _first: CellContext = _
+  @volatile private var _last: CellContext = _
+
+  def insert(cellContext: CellContext, after: Option[CellID]): CellContext = after match {
+    case None => insertFirst(cellContext)
+    case Some(id) if id == cellContext.id => throw new IllegalArgumentException("Cell can not be inserted after itself")
+    case Some(id) =>
+      @tailrec def findTail(at: Option[CellContext], currentTail: List[CellContext]): List[CellContext] = at match {
+        case Some(ctx) if ctx.id == id => ctx :: currentTail
+        case Some(ctx) => findTail(ctx.previous, ctx :: currentTail)
+        case None => Nil
+      }
+
+      synchronized {
+        findTail(Option(_last), Nil) match {
+          case Nil => throw new NoSuchElementException(s"Predecessor $id not found")
+          case at :: tail =>
+            val cellsBefore = at.collectBack {
+              case ctx => ctx.id
+            }.toSet
+
+            cellContext.setPrev(at)
+
+            tail.headOption.foreach(_.setPrev(cellContext))
+
+            tail.filter(_.previous.exists(ctx => cellsBefore(ctx.id))).foreach {
+              ctx => ctx.setPrev(cellContext)
+            }
```

Comment:
hm, having a bit of trouble figuring out what's going on here.

Reply:
You're right, I think this is unnecessary. The point was to make sure that if any cell after the insertion point was pointing at a cell before the insertion point, to re-point it. But that can't really happen. So I deleted this code in #191.